rpc/RPCTransaction.java

```diff
@@ -139,10 +130,11 @@ public void rollback() {
 
     @Override
     public void close() {
-        if (streamIsOpen.compareAndSet(true, false)) {
+        if (transactionWasClosed.compareAndSet(false, true)) {
```

Comment:
We use our `AtomicBoolean` to ensure `collectors` gets purged once and only once, and that after the purge has started, no more collectors can be added to it.

The purge flushes all the collectors' `BlockingQueue`s by sending them an error response. This, in turn, causes all `QueryFutures` and `Stream`s to immediately throw `TRANSACTION_CLOSED`.

Note that `close` can be triggered either by the client calling `close`, or the server terminating the transaction stream (e.g. after a commit)